subworkflows/local/utils_nfcore_nallo_pipeline/main.nf

```diff
@@ -506,6 +334,52 @@ def methodsDescriptionText(mqc_methods_yaml) {
     return description_html.toString()
 }
 
+def extractSoftwareFromVersions(module_yaml_file) {
+    def yaml = new org.yaml.snakeyaml.Yaml()
+    def yamlData = yaml.load(module_yaml_file)
+    // Extract all software (keys) from a module yaml
+    def softwareInModule = yamlData.values().collect { it.keySet() }.flatten()
+    return softwareInModule
+}
+
+def generateReferenceHTML(tool_list, description) {
+    def items = tool_list
+        .collect { it.trim() }
+        .unique()              // samtools and bcftools share reference
+        .findAll { it != "" }  // some tools does not have a reference, e.g. awk, gunzip
+
+    if (description == 'citation') {
+        return "  <p>Tools used in the workflow included: ${items.join(', ')}.</p>"
+    } else if (description == 'bibliography') {
+        return "  <h4>References</h4><ul><li>${items.join('</li><li>')}</li></ul>"
+    }
+}
+
+def citationBibliographyText(ch_versions, references_yaml, description) {
+    def yaml = new org.yaml.snakeyaml.Yaml()
+    def softwareReferences = yaml.load(references_yaml.text).tool
+
+    def unwantedReferences = ['genomic-medicine-sweden/nallo', 'Nextflow']
+    // These are not collected in ch_versions but should be referenced
+    baseTools = Channel.from(['nextflow', 'nf_core', 'bioconda', 'biocontainers', 'multiqc'])
+
+    ch_versions
+        .map { module_yaml -> extractSoftwareFromVersions(module_yaml) }
+        .flatten() // split multi-tool modules
+        .unique()
+        .filter { tool -> !unwantedReferences.contains(tool) }
+        .concat(baseTools)
+        .collect { tool ->
+            def toolDetails = softwareReferences[tool]
+            if (toolDetails == null) {
+                throw new IllegalStateException("Tool: '${tool}' not found in ${references_yaml}")
+            }
```

Comment:
Is this to make sure that each tool we run has an entry in the `software_reference.yml` file, even if the values are empty? Should we add this to the contribution doc?

Reply:
This is both to make sure we have an entry, but also because `toolDetails[description]` below will error with a less informative "can't get null value" or similar..

Good idea, will add to contribution doc!